src/VectorizeLoops.cpp

```diff
@@ -10,23 +10,209 @@
 #include "IREquality.h"
 #include "ExprUsesVar.h"
 #include "Solve.h"
+#include "Simplify.h"
+#include "CSE.h"
 
 namespace Halide {
 namespace Internal {
 
 using std::string;
 using std::vector;
+using std::pair;
+
+// For a given var, replace expressions like shuffle_vector(var, 4)
```

Comment:
Minor comment: most of the code in files like this should be in an anonymous namespace. I've been adding these when it's convenient and I remember it, but I think we should be doing it anyways. I'm pointing this out here because there's lots of new stuff in this file.